### PR TITLE
Range Enrich

### PR DESCRIPTION
**Pull request by rafaelgmota**
Implementation of range methods: contains, equals, includes, overlaps and abuts. 
The implementation of combination and isContiguous are missing and will be done.

### CODE REVIEW

File: src/util/Range.java [hunk 1]
```diff
@@ -34,4 +34,66 @@ public int max(int dimension) {
         assert dimension < dimensions();
         return this.max[dimension];
     }
+
+    public boolean contains(int x) {
+        if(numberOfDimensions > 1)
+            return false;
+        if(x < min[0] || x > max[0])
```
Comment (jmferreiratech) [inline src/util/Range.java]: These three-line portion could be simplified to an one-line return.

File: src/util/Range.java [hunk 2]
```diff
@@ -34,4 +34,66 @@ public int max(int dimension) {
...
+            return false;
+        return true;
+    }
+
+    public boolean contains(int x, int y) {
+        if (numberOfDimensions < 2 || numberOfDimensions > 2)
```
Comment (jmferreiratech) [inline src/util/Range.java]: It could check like this: "numberOfDimensions != 2"

File: src/util/Range.java [hunk 3]
```diff
@@ -34,4 +34,66 @@ public int max(int dimension) {
...
+            return false;
+
+            if(x < min[0] || x > max[0] || y < min[1] || y > max[1])
```
Comment (jmferreiratech) [inline src/util/Range.java]: These three-line portion could be simplified to an one-line return

File: tests/RangeTest.java [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+        Assert.assertFalse(range1.contains(21, 20));
+    }
+
+    @Test
+    public void OneDimensionEquals() throws Exception {
+        Range range1 = Range.OneDimensionalRange(0, 20);
+        Range range2;
+        //Equals
+        range2 = Range.OneDimensionalRange(0, 20);
+        Assert.assertTrue(range1.equals(range2));
```
Comment (jmferreiratech) [inline tests/RangeTest.java]: You could use Assert.assertEquals here. It'll call your overrided equals method.

File: tests/RangeTest.java [hunk 5]
```diff
@@ -0,0 +1,157 @@
+import org.junit.*;
+import util.*;
+
+public class RangeTest {
```
Comment (jmferreiratech) [inline tests/RangeTest.java]: Each comment you put inside a unit test seems to indicate the start of a new unit test and so you should break them to smaller unit tests with one meaning each.

File: tests/RangeTest.java [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+
+    @Test
```
Comment (jmferreiratech) [inline tests/RangeTest.java]: You could show first all unit tests for OneDimensionalRanges then those to TwoDimensionalRanges

File: src/util/Range.java [hunk 7]
```diff
@@ -34,4 +34,66 @@ public int max(int dimension) {
...
+        else {
+            for(int i = 0; i < numberOfDimensions; i++) {
+                if(other.min[i] != min[i] || other.max[i] != max[i])
+                    return false;
+            }
+            return true;
+        }
+    }
+
+    public boolean includes(Range arg) {
```
Comment (jmferreiratech) [inline src/util/Range.java]: We like to call same class objects passed as parameter, "that" to contrast with "this" and in this case is more uniform to use the "this" specifier and make the same kind of attribute access (directly or by accessors).

File: src/util/Range.java [hunk 8]
```diff
@@ -34,4 +34,66 @@ public int max(int dimension) {
...
+    @Override
+    public boolean equals(Object arg) {
+        if(!(arg instanceof Range))
+            return false;
+
+        Range other = (Range)arg;
+        if(other.dimensions() != numberOfDimensions) {
+            return false;
+        }
+        else {
```
Comment (jmferreiratech) [inline src/util/Range.java]: We don't need else here.

File: src/util/Range.java [hunk 9]
```diff
@@ -34,4 +34,66 @@ public int max(int dimension) {
...
+            for(int i = 0; i < numberOfDimensions; i++) {
+                if(other.min[i] != min[i] || other.max[i] != max[i])
+                    return false;
+            }
+            return true;
+        }
+    }
+
+    public boolean includes(Range arg) {
+        for(int i = 0; i < numberOfDimensions; i++) {
```
Comment (jmferreiratech) [inline src/util/Range.java:87]: Take care with the number of dimensions of parameters.

File: src/util/Range.java [hunk 10]
```diff
@@ -34,4 +34,66 @@ public int max(int dimension) {
...
+                return false;
+        }
+        return true;
+    }
+
+    public boolean abuts(Range arg) {
+        if (overlaps(arg))
+            return false;
+
+        for(int i = 0 ; i < numberOfDimensions; i++)
```
Comment (jmferreiratech) [inline src/util/Range.java]: Make more uniform use of curly braces for one statement blocks. There's no explicit convention but we'd prefer always put curly braces.

File: src/util/Range.java [hunk 11]
```diff
@@ -34,4 +34,66 @@ public int max(int dimension) {
...
+            if(arg.min(i) > this.max[i]+1)
+                return false;
+
```
Comment (jmferreiratech) [inline src/util/Range.java:128]: Delete that line.

**Review by jmferreiratech — CHANGES_REQUESTED** [review]
Small changes. Only the unit tests will require more of your work.

**Comment by jmferreiratech** [comment]
Please put the issue number on the commit message.

File: src/util/Range.java (hunk 1 above)
Comment (rafaelgmota) [reply src/util/Range.java]: Done

File: src/util/Range.java (hunk 2 above)
Comment (rafaelgmota) [reply src/util/Range.java]: Done

File: src/util/Range.java (hunk 3 above)
Comment (rafaelgmota) [reply src/util/Range.java]: Done

File: src/util/Range.java (hunk 8 above)
Comment (rafaelgmota) [reply src/util/Range.java]: Done

File: src/util/Range.java (hunk 7 above)
Comment (rafaelgmota) [reply src/util/Range.java]: Done

File: src/util/Range.java (hunk 10 above)
Comment (rafaelgmota) [reply src/util/Range.java]: Done

File: src/util/Range.java (hunk 11 above)
Comment (rafaelgmota) [reply src/util/Range.java:128]: Done

File: src/util/Range.java (hunk 9 above)
Comment (rafaelgmota) [reply src/util/Range.java:87]: Done

**Comment by rafaelgmota** [comment]
RangeTest was split into two files, one for one dimension range and another to two dimension ones.
The above improvements were done in both files.

File: src/util/Range.java [hunk 12]
```diff
@@ -34,4 +34,75 @@ public int max(int dimension) {
         assert dimension < dimensions();
         return this.max[dimension];
     }
+
+    public boolean contains(int x) {
+        if(numberOfDimensions != 1 || x < min[0] || x > max[0]) {
+            return false;
+        }
+        return true;
```
Comment (jmferreiratech) [inline src/util/Range.java]: I think that is more clear if you use "return (min[0] <= x && x <= max[0]);"

File: src/util/Range.java [hunk 13]
```diff
@@ -34,4 +34,75 @@ public int max(int dimension) {
...
+    }
+
+    public boolean contains(int x, int y) {
+        if (numberOfDimensions != 2 || x < min[0] || x > max[0] || y < min[1] || y > max[1]) {
+            return false;
+        }
+        return true;
```
Comment (jmferreiratech) [inline src/util/Range.java]: Same idea as before.

File: src/util/Range.java [hunk 14]
```diff
@@ -34,4 +34,75 @@ public int max(int dimension) {
...
+    }
+
+    @Override
+    public boolean equals(Object arg) {
+        if(!(arg instanceof Range)) {
+            return false;
+        }
+
+        Range other = (Range)arg;
```
Comment (jmferreiratech) [inline src/util/Range.java]: Name it "that" and use "this" to contrast.

File: src/util/Range.java [hunk 15]
```diff
@@ -34,4 +34,75 @@ public int max(int dimension) {
...
+    }
+
+    public boolean abuts(Range that) {
+        if(that.dimensions() != this.dimensions()) {
+            return false;
+        }
+        if (this.overlaps(that)) {
+            return false;
+        }
+        for(int i = 0 ; i < numberOfDimensions; i++) {
```
Comment (jmferreiratech) [inline src/util/Range.java]: remove the space after "int i = 0"

**Review by jmferreiratech — CHANGES_REQUESTED** [review]
More improvements. Thanks

**Comment by rafaelgmota** [comment]
Implementation of isContiguous, only valid for one and two dimensions, need to be extended for three dimensions.

**Comment by rafaelgmota** [comment]
Implementation of combination Range method

**Comment by rafaelgmota** [comment]
Using Range functions at Region.java and RBR.java

**Comment by rafaelgmota** [comment]
Refactoring of Range.combination(). 
Using Math.min() and Math.max()

File: src/util/Range.java [hunk 16]
```diff
@@ -15,6 +17,12 @@ public static Range TwoDimensionalRange(int xMin, int xMax, int yMin, int yMax)
         return new Range(2, new int[]{xMin, yMin}, new int[]{xMax, yMax});
     }
 
+    private Range(int[]min, int[]max) {
```
Comment (jmferreiratech) [inline src/util/Range.java]: You don't need it.

File: src/util/Range.java [hunk 17]
```diff
@@ -34,4 +42,103 @@ public int max(int dimension) {
...
+            return false;
+        }
+        return (min[0] <= x && x <= max[0]);
+    }
+
+    public boolean contains(int x, int y) {
+        if (numberOfDimensions != 2) {
+            return false;
+        }
+        return (x >= min[0]&&x <= max[0]&&y >= min[1]&&y <= max[1]);
```
Comment (jmferreiratech) [inline src/util/Range.java]: Put spaces around operators.

File: src/util/Range.java [hunk 18]
```diff
@@ -34,4 +42,103 @@ public int max(int dimension) {
...
+        }
+        Range that = (Range)arg;
+        if(that.dimensions() != numberOfDimensions) {
+            return false;
+        }
+        for(int i = 0; i < numberOfDimensions; i++) {
+              if(that.min(i) != this.min(i) || that.max(i) != this.max(i)) {
+                  return false;
+               }
+           }
```
Comment (jmferreiratech) [inline src/util/Range.java]: Align this close brace.

File: src/util/Range.java [hunk 19]
```diff
@@ -34,4 +42,103 @@ public int max(int dimension) {
...
+        return true;
+    }
+
+    public boolean includes(Range that) {
+        if(that.dimensions() != this.dimensions()) {
+            return false;
+        }
+        for(int i = 0; i < numberOfDimensions; i++) {
+            if(that.min(i) < this.min(i) || that.max(i) > this.max(i)) {
```
Comment (jmferreiratech) [inline src/util/Range.java:76]: Couldn't you use the contains() method here?

File: src/util/Range.java [hunk 20]
```diff
@@ -34,4 +42,103 @@ public int max(int dimension) {
...
+        }
+        return true;
+    }
+
+    public boolean overlaps(Range that) {
+        if(that.dimensions() != this.dimensions()) {
+            return false;
+        }
+        for(int i = 0; i < numberOfDimensions; i++) {
+            if(that.max(i) < this.min(i) || that.min(i) > this.max(i)) {
```
Comment (jmferreiratech) [inline src/util/Range.java:88]: Couldn't you use the contains() method here also?

**Review by jmferreiratech — CHANGES_REQUESTED** [review]
Fix these points and I'll merge it. Thanks

File: src/util/Range.java (hunk 19 above)
Comment (rafaelgmota) [reply src/util/Range.java:76]: The above logic is valid for one and two dimensions Range and we have one contains() for each dimension. So, will be necessary some if/else to treat each dimension diferently.

File: src/util/Range.java (hunk 20 above)
Comment (rafaelgmota) [reply src/util/Range.java:88]: Same as above

**Comment by rafaelgmota** [comment]
fix last comments

**Review by jmferreiratech — APPROVED** [review]

**Pull request merged**